src/include/optimizer/abstract_node.h

```diff
@@ -0,0 +1,129 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// abstract_node.h
+//
+// Identification: src/include/optimizer/abstract_node.h
+//
+// Copyright (c) 2015-16, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "optimizer/property_set.h"
+#include "util/hash_util.h"
+
+#include <memory>
+#include <string>
+
+namespace peloton {
+namespace optimizer {
+
+enum class OpType {
+  Undefined = 0,
+  // Special match operators
+  Leaf,
+  // Logical ops
+  Get,
+  LogicalExternalFileGet,
+  LogicalQueryDerivedGet,
+  LogicalProjection,
+  LogicalFilter,
+  LogicalMarkJoin,
+  LogicalDependentJoin,
+  LogicalSingleJoin,
+  InnerJoin,
+  LeftJoin,
+  RightJoin,
+  OuterJoin,
+  SemiJoin,
+  LogicalAggregateAndGroupBy,
+  LogicalInsert,
+  LogicalInsertSelect,
+  LogicalDelete,
+  LogicalUpdate,
+  LogicalLimit,
+  LogicalDistinct,
+  LogicalExportExternalFile,
+  // Separate between logical and physical ops
+  LogicalPhysicalDelimiter,
+  // Physical ops
+  DummyScan, /* Dummy Physical Op for SELECT without FROM*/
+  SeqScan,
+  IndexScan,
+  ExternalFileScan,
+  QueryDerivedScan,
+  OrderBy,
+  PhysicalLimit,
+  Distinct,
+  InnerNLJoin,
+  LeftNLJoin,
+  RightNLJoin,
+  OuterNLJoin,
+  InnerHashJoin,
+  LeftHashJoin,
+  RightHashJoin,
+  OuterHashJoin,
+  Insert,
+  InsertSelect,
+  Delete,
+  Update,
+  Aggregate,
+  HashGroupBy,
+  SortGroupBy,
+  ExportExternalFile,
+};
+
+//===--------------------------------------------------------------------===//
+// Abstract Node
+//===--------------------------------------------------------------------===//
+//TODO(ncx): dependence on OperatorVisitor
+class OperatorVisitor;
+
+struct AbstractNode {
+  AbstractNode() {}
+  AbstractNode(std::shared_ptr<AbstractNode> node) : node(node) {}
+
+  ~AbstractNode() {}
+
+  virtual void Accept(OperatorVisitor *v) const = 0;
```

Comment:
Documentation on this would be really useful

Reply:
Resolved.